runtime/extern.c

```diff
@@ -1130,3 +1130,66 @@ CAMLexport void caml_serialize_block_float_8(void * data, intnat len)
   }
 #endif
 }
+
+CAMLprim value caml_obj_reachable_words(value v)
+{
+  intnat size;
+  struct extern_item * sp;
+  uintnat h = 0;
+  uintnat pos;
+
+  extern_init_position_table();
+  sp = extern_stack;
+  size = 0;
+  while (1) {
+    if (Is_long(v)) {
+      /* Tagged integers contribute 0 to the size, nothing to do */
+#ifndef NO_NAKED_POINTERS
+    } else if (! Is_in_heap_or_young(v)) {
+      /* Out-of-heap blocks contribute 0 to the size, nothing to do */
+      /* However, once we get rid of the page table, we will no longer
+         be able to distinguish major heap blocks and out-of-heap blocks,
+         so we will need to count out-of-heap blocks too. */
+#endif
+    } else if (extern_lookup_position(v, &pos, &h)) {
+      /* Already seen and counted, nothing to do */
+    } else {
+      header_t hd = Hd_val(v);
+      tag_t tag = Tag_hd(hd);
+      mlsize_t sz = Wosize_hd(hd);
+      /* Infix pointer: go back to containing closure */
+      if (tag == Infix_tag) {
+        v = v - Infix_offset_hd(hd);
```

Comment:
Should we check if v was already visited here ?

Reply:
No need to.  I chose to enter the closure itself in the hash table, but not its infix pointers.  Sharing will be detected and recorded after we go back to the containing closure.

Reply:
Right, that make sense.